pyodide-build/pyodide_build/build_env.py

```diff
@@ -0,0 +1,230 @@
+
+    tools_dir = Path(__file__).parent / "tools"
+
+    env["CMAKE_TOOLCHAIN_FILE"] = str(
+        tools_dir / "cmake/Modules/Platform/Emscripten.cmake"
+    )
+    env["PYO3_CONFIG_FILE"] = str(tools_dir / "pyo3_config.ini")
+
+    hostsitepackages = env["HOSTSITEPACKAGES"]
+    pythonpath = [
+        hostsitepackages,
+    ]
+    env["PYTHONPATH"] = ":".join(pythonpath)
+    return env
+
+
+def get_build_flag(name: str) -> str:
+    """
+    Get a value of a build flag.
+    """
+    build_vars = get_build_environment_vars()
+    if name not in build_vars:
+        raise ValueError(f"Unknown build flag: {name}")
+
+    return build_vars[name]
+
+
+def _set_pyodide_root(*, quiet: bool = False) -> None:
+    """
+    Initialize Pyodide build environment, namely set PYODIDE_ROOT environment variable.
+
+    This function works both in-tree and out-of-tree builds:
+    - In-tree builds: Searches for the root of the Pyodide repository in parent directories
+    - Out-of-tree builds: Downloads and installs the Pyodide build environment into the current directory
+
+    Parameters
+    ----------
+    quiet
+        If True, do not print any messages
+    """
+
+    from . import install_xbuildenv  # avoid circular import
+
+    # If we are building docs, we don't need to know the PYODIDE_ROOT
+    if "sphinx" in sys.modules:
+        os.environ["PYODIDE_ROOT"] = ""
+        return
+
+    # 1) If PYODIDE_ROOT is already set, do nothing
+    if "PYODIDE_ROOT" in os.environ:
+        return
+
+    # 2) If we are doing an in-tree build,
+    #    set PYODIDE_ROOT to the root of the Pyodide repository
+    try:
+        os.environ["PYODIDE_ROOT"] = str(search_pyodide_root(Path.cwd()))
+        return
+    except FileNotFoundError:
+        pass
+
+    # 3) If we are doing an out-of-tree build,
+    #    download and install the Pyodide build environment
+    xbuildenv_path = Path(".pyodide-xbuildenv").resolve()
+
+    if xbuildenv_path.exists():
+        os.environ["PYODIDE_ROOT"] = str(xbuildenv_path / "xbuildenv" / "pyodide-root")
+        return
+
+    with ExitStack() as stack:
+        if quiet:
+            # Prevent writes to stdout
+            stack.enter_context(redirect_stdout(StringIO()))
+
+        # install_xbuildenv will set PYODIDE_ROOT env variable, so we don't need to do it here
+        # TODO: return the path to the xbuildenv instead of setting the env variable inside install_xbuildenv
+        install_xbuildenv.install(xbuildenv_path, download=True)
+
+
+@functools.cache
+def _get_make_environment_vars() -> dict[str, str]:
+    """
+    Load environment variables from Makefile.envs
+
+    This is not a public API, use get_build_environment_vars instead.
+    """
+
+    from .common import exit_with_stdio  # avoid circular import
+
+    PYODIDE_ROOT = get_pyodide_root()
+    environment = {}
+    result = subprocess.run(
+        ["make", "-f", str(PYODIDE_ROOT / "Makefile.envs"), ".output_vars"],
```

Comment:
Unrelated to this PR: WDYT if we move part of the configuration intended to be modifiable by the user from `Makefile.envs` to some toml file and have some script that sets those env variables if needed. Not in this PR. Just to try to move away from makefiles?

Reply:
I think, for now, setting env variables e.g. `WASM_LIBRARY_DIR=... pyodide build recipes` will overwrite Makefile.envs.

But yes, I do want to move some of the configurations in Makefile.envs  that are not used in building the main module to somewhere else that users can easily modify.